### DISCUSSION TITLE
W3 validator throws error but g.parse() not

### DISCUSSION THREAD

**Discussion by Delilovic** [RDFLib/rdflib#1712] [Q&A] [answered]
Hello I have a simple XML file with the following content:

```xml
<?xml version="1.0" encoding="UTF-8"?>
<root>
    <child>OKOK</child>
</root>
```

When checking the content with the w3 Validation Tool https://www.w3.org/RDF/Validator/ I get an error:
Error: Your document does not contain any RDF statement. But when I test it with rdflib, it parses the file without error.

Do you know why rdflib does not throw an error message? And if rdflib is somehow able to recognize that the RDF tags are missing in the file?

Kind Regards

**Comment by ghost** [comment, marked as answer]
> Do you know why rdflib does not throw an error message?

That's a reasonable question. RDFLib adopts the widely-accepted stance that a library is more useful to programmers if it is tolerant about what it accepts but strict about what it produces. In this instance, RDFLib is being useful, allowing us to work with the input.

Here's how it might go. Let's start by parsing the data:
```pycon
>>> data = """<?xml version="1.0" encoding="UTF-8"?><root><child>OKOK</child></root>"""
>>> g = Graph()
>>> g.bind("ex", Namespace("urn:example:"))
>>> g.parse(data=data, format="xml")
```
As you observed, this does not produce an error and this is because the result of the parse is now something we can work with ...
```pycon
>>> for triple in g.triples((None, None, None)):
...    print(triple)
(rdflib.term.BNode('N44cccf29cf14487fa1833fb865af4531'),  rdflib.term.URIRef('http://www.w3.org/1999/02/22-rdf-syntax-ns#type'),  rdflib.term.URIRef('root'))
(rdflib.term.BNode('N44cccf29cf14487fa1833fb865af4531'), rdflib.term.URIRef('child'), rdflib.term.Literal('OKOK'))
```
Taking it a wee bit further with a rough'n'ready function to add a namespace where appropriate:
```py
>>> def addns(x):
>>>     return URIRef('urn:example:' + x) if isinstance(x, URIRef) and not x.startswith('http') else x
```
Gives us the opportunity to do this:
```py
>>> for (s, p, o) in g.triples((None, None, None)):
...     g.remove((s, p, o))
....    g.add((s, addns(p), addns(o)))
```
Bringing us to the point where we can copy'n'paste the serialized graph into the W3C validator for a successful validation:
```py
print(g.serialize(format="xml"))
```
```xml
<?xml version="1.0" encoding="utf-8"?>
<rdf:RDF
   xmlns:ex="urn:example:"
   xmlns:rdf="http://www.w3.org/1999/02/22-rdf-syntax-ns#"
>
  <rdf:Description rdf:nodeID="N44cccf29cf14487fa1833fb865af4531">
    <rdf:type rdf:resource="urn:example:root"/>
    <ex:child>OKOK</ex:child>
  </rdf:Description>
</rdf:RDF>
```

> And if rdflib is somehow able to recognize that the RDF tags are missing in the file?

It has done ... and has tried to fix things up a bit with the intention that the result is usefully computationally tractable.

HTH

**Comment by Delilovic** [comment]
Thank you @gjhiggins for the detailed explanation. I also think that flexibility is an advantage. Only I do not understand why it is not possible to catch a warning. Attached is an example from Jena (java implementation) the first two examples give no error but a warning the third example gives an error and also a warning.

Would such a warning also be useful with rdflib or have I overlooked this feature?

Greetings

**Example 1 (Valid but has warnings)** 

```xml
<?xml version="1.0" encoding="UTF-8"?>
<root>
    <child>OK</child>
</root>
```

**Log 1**

```
org.apache.jena.riot:101 - [line: 1, col: 45] {W104} Unqualified typed nodes are not allowed. Type treated as a relative URI.
org.apache.jena.riot:101 - [line: 1, col: 45] {W136} Relative URIs are not permitted in RDF: specifically <root>
org.apache.jena.riot:101 - [line: 2, col: 12] {W104} Unqualified property elements are not allowed. Treated as a relative URI.
org.apache.jena.riot:101 - [line: 2, col: 12] {W136} Relative URIs are not permitted in RDF: specifically <child>
```

**Example 2 (Valid but has warnings)** 

```xml
<?xml version="1.0" encoding="UTF-8"?>
<root>
    <child>OK</child>
    <child>OK</child>
</root>
```

**Log 2**

```
org.apache.jena.riot:101 - [line: 1, col: 45] {W104} Unqualified typed nodes are not allowed. Type treated as a relative URI.
org.apache.jena.riot:101 - [line: 1, col: 45] {W136} Relative URIs are not permitted in RDF: specifically <root>
org.apache.jena.riot:101 - [line: 2, col: 12] {W104} Unqualified property elements are not allowed. Treated as a relative URI.
org.apache.jena.riot:101 - [line: 2, col: 12] {W136} Relative URIs are not permitted in RDF: specifically <child>
org.apache.jena.riot:101 - [line: 3, col: 12] {W104} Unqualified property elements are not allowed. Treated as a relative URI.
org.apache.jena.riot:101 - [line: 3, col: 12] {W136} Relative URIs are not permitted in RDF: specifically <child>
```

**Example 3 (Invalid and has warnings)** 

```xml
<?xml version="1.0" encoding="UTF-8"?>
<root>
    <child>
        <subchild>NOT OK</subchild>
        <subchild>NOT OK</subchild>
    </child>
</root>
```

**Log 3**

```
org.apache.jena.riot:101 - [line: 1, col: 45] {W104} Unqualified typed nodes are not allowed. Type treated as a relative URI.
org.apache.jena.riot:101 - [line: 1, col: 45] {W136} Relative URIs are not permitted in RDF: specifically <root>
org.apache.jena.riot:101 - [line: 2, col: 12] {W104} Unqualified property elements are not allowed. Treated as a relative URI.
org.apache.jena.riot:101 - [line: 2, col: 12] {W136} Relative URIs are not permitted in RDF: specifically <child>
org.apache.jena.riot:101 - [line: 3, col: 19] {W104} Unqualified typed nodes are not allowed. Type treated as a relative URI.
org.apache.jena.riot:101 - [line: 3, col: 19] {W136} Relative URIs are not permitted in RDF: specifically <subchild>
org.apache.jena.riot:107 - [line: 3, col: 27] {E202} Expecting XML start or end element(s). String data "NOT OK" not allowed. Maybe there should be an rdf:parseType='Literal' for embedding mixed XML content in RDF. Maybe a striping error.
Expecting XML start or end element(s). String data "NOT OK" not allowed. Maybe there should be an rdf:parseType='Literal' for embedding mixed XML content in RDF. Maybe a striping error.
```

Kind Regards

**Comment by ghost** [comment]
> I do not understand why it is not possible to catch a warning.

It's certainly _possible_, the code for the rdfxml parser is [here](https://github.com/RDFLib/rdflib/blob/6ed2ef48ed38679bcdafe7cae250a2ef4b315e7b/rdflib/plugins/parsers/rdfxml.py) and anyone who is sufficiently both motivated and skilled could modify the parser to emit such a warning and submit a PR. However, given the [venerability](https://github.com/RDFLib/rdflib/blob/d7850e3cb00b5e5f1aca9445fc754a346bbcb301/rdflib/syntax/parsers/RDFXMLHandler.py) of that code (“Copyright (c) 2002, Daniel Krech, http://eikeon.com/”), the lack of a warning has not been perceived either as a pressing issue or as a critical feature so is unlikely to figure on anyone's to-do list - which is not surprising if you understand it as an issue of allocating scarce resources.

**Comment by Delilovic** [comment]
> which is not surprising if you understand it as an issue of allocating scarce resources


Understandable, the library is already in amazing condition, these are just cosmetic details.

I wish you all the best and thank you for your efforts.

**Comment by ghost** [comment]
Thanks for your understanding, I'll migrate this to the Discussions Q&A section where it'll remain more accessible than a closed issue.
